dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/jdbc/JdbcCustomIconStore.java

```diff
@@ -0,0 +1,138 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.icon.jdbc;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.hisp.dhis.fileresource.FileResource;
+import org.hisp.dhis.icon.CustomIcon;
+import org.hisp.dhis.icon.CustomIconStore;
+import org.hisp.dhis.user.User;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowCallbackHandler;
+import org.springframework.stereotype.Repository;
+
+@Slf4j
+@Repository( "org.hisp.dhis.icon.CustomIconStore" )
+@RequiredArgsConstructor
+public class JdbcCustomIconStore implements CustomIconStore
+{
+    private final JdbcTemplate jdbcTemplate;
+
+    @Override
+    public CustomIcon getIconByKey(String key) {
+        List<CustomIcon> customIcons = new ArrayList<>();
+
+        final String sql = """
+                    select c.key as iconkey, c.description as icondescription, c.keywords as keywords, f.uid as fileresourceuid, u.uid as useruid
+                    from customicon c join fileresource f on f.fileresourceid = c.fileresourceid
+                    join userinfo u on u.userinfoid = c.createdby
+                    where key = ?
+                    """;
+
+        jdbcTemplate.query(sql, getRowCallbackHandler(customIcons), key);
+
+        return customIcons.isEmpty() ? null : customIcons.get(0);
+    }
+
+    @Override
+    public List<CustomIcon> getIconsByKeywords(String[] keywords) {
+        List<CustomIcon> customIcons = new ArrayList<>();
+
+        final String sql = """
+                    select c.key as iconkey, c.description as icondescription, c.keywords as keywords, f.uid as fileresourceuid, u.uid as useruid
+                    from customicon c join fileresource f on f.fileresourceid = c.fileresourceid
+                    join userinfo u on u.userinfoid = c.createdby
+                    where keywords @> string_to_array(?,',')
+                    """;
+
+        jdbcTemplate.query(sql, getRowCallbackHandler(customIcons), String.join(",", keywords));
```

Comment:
Have you tried https://www.baeldung.com/spring-jdbctemplate-in-list#NamedParameterJdbcTemplate so we do not have to join in Java and split again in SQL?